policy/scan/pdque/queue.go

```diff
@@ -0,0 +1,414 @@
+// Close safely shuts down the queue, ensuring all resources are released.
+func (q *Queue) Close() error {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	// If the queue is already closed, return an error or just exit.
+	if q.closed {
+		return ErrQueueClosed
+	}
+
+	// Clean up temporary files.
+	files, err := os.ReadDir(q.path)
+	if err != nil {
+		return err
+	}
+	for _, file := range files {
+		if strings.HasPrefix(file.Name(), tempFilePrefix) {
+			err := os.Remove(filepath.Join(q.path, file.Name()))
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Set the queue as closed to prevent further operations.
+	q.closed = true
+
+	// Wake up all goroutines waiting on the condition variable before closing the queue.
+	q.cond.Broadcast()
+
+	return nil
+}
+
+func (q *Queue) Enqueue(obj interface{}) error {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	if q.closed {
+		return ErrQueueClosed
+	}
+
+	// Check if the queue size has reached the maxSize
+	if q.maxSize > 0 {
+		size, err := q.currentSize()
+		if err != nil {
+			return err // Handle or return the error
+		}
+		if size >= q.maxSize {
+			return ErrQueueFull // Or handle as needed
+		}
+	}
+
+	// Find the next available filename
+	filename, err := q.nextAvailableFilename()
+	if err != nil {
+		return err
+	}
+
+	tempPath := filepath.Join(q.path, tempFilePrefix+filename+tempFileExt)
+	finalPath := filepath.Join(q.path, filename+jobFileExt)
+
+	// Encode the struct to a byte buffer
+	var buff bytes.Buffer
+	enc := gob.NewEncoder(&buff)
+	if err := enc.Encode(obj); err != nil {
+		return err
+	}
+
+	// Write to a temporary file
+	err = os.WriteFile(tempPath, buff.Bytes(), 0o644)
+	if err != nil {
+		return err
+	}
+
+	// Rename the temporary file to its final name
+	err = os.Rename(tempPath, finalPath)
+	if err != nil {
+		return err
+	}
+
+	// After successfully enqueueing a job, wake up one of the waiting goroutines, if any.
+	q.cond.Broadcast()
+
+	return nil
+}
+
+func (q *Queue) Dequeue() (interface{}, error) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	if q.closed {
+		return nil, ErrQueueClosed
+	}
+
+	files, err := os.ReadDir(q.path)
+	if err != nil {
+		return nil, err
+	}
+
+	// Sort job files by name (which is the timestamp)
+	sort.Slice(files, func(i, j int) bool {
+		return files[i].Name() < files[j].Name()
+	})
+
+	for _, file := range files {
+		if filepath.Ext(file.Name()) == jobFileExt {
+			jobPath := filepath.Join(q.path, file.Name())
+
+			data, err := os.ReadFile(jobPath)
+			if err != nil {
+				return nil, err
+			}
+
+			// Decode the bytes into an object
+			obj := q.builder()
+			if err := gob.NewDecoder(bytes.NewReader(data)).Decode(obj); err != nil {
+				return nil, ErrUnableToDecode{
+					Path: q.path,
+					Err:  err,
+				}
+			}
+
+			// Remove the job file
+			if err := os.Remove(jobPath); err != nil {
+				return nil, err
+			}
+
+			return obj, nil
+		}
+	}
+
+	return nil, errors.New("no jobs in queue")
+}
+
+//func (q *Queue) DequeueBlock() (interface{}, error) {
```

Comment:
Are we keeping this around to enable it shortly after merging, or is this a bit of a leftover?
If it's something that comes in a quick follow-up, I'd be ok to leave it (as an exception). Otherwise, generally, it'd be great to keep the code clean :)
(I suspect it's by error since the same method name is defined below)